src/Components/Analyzers/src/ComponentParameterUsageAnalyzer.cs

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.Operations;
+
+namespace Microsoft.AspNetCore.Components.Analyzers
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class ComponentParameterUsageAnalyzer : DiagnosticAnalyzer
+    {
+        public ComponentParameterUsageAnalyzer()
+        {
+            SupportedDiagnostics = ImmutableArray.Create(new[]
+            {
+                DiagnosticDescriptors.ComponentParametersShouldNotBeSetOutsideOfComponents,
+            });
+        }
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; }
+
+        public override void Initialize(AnalysisContext context)
+        {
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics);
+            context.RegisterCompilationStartAction(context =>
+            {
+                if (!ComponentSymbols.TryCreate(context.Compilation, out var symbols))
+                {
+                    // Types we need are not defined.
+                    return;
+                }
+
+                context.RegisterOperationBlockStartAction(startBlockContext =>
+                {
+                    var assignedPropertyReferences = new List<IPropertyReferenceOperation>();
+                    startBlockContext.RegisterOperationAction(context =>
+                    {
+                        foreach (var operation in context.Operation.Children)
+                        {
+                            if (operation.Kind == OperationKind.PropertyReference)
+                            {
+                                assignedPropertyReferences.Add((IPropertyReferenceOperation)operation);
+                            }
+                        }
+                    }, OperationKind.SimpleAssignment, OperationKind.CompoundAssignment, OperationKind.CoalesceAssignment);
+
+                    startBlockContext.RegisterOperationBlockEndAction(context =>
+                    {
+                        for (var i = 0; i < assignedPropertyReferences.Count; i++)
+                        {
+                            var propertyReference = assignedPropertyReferences[i];
+                            var componentProperty = propertyReference.Member;
+                            var propertyAttributes = componentProperty.GetAttributes();
+
+                            if (!ComponentFacts.HasParameterAttribute(symbols, propertyAttributes))
```

Comment:
Is there a reason why we can't use `IsParameter` here instead? It seems like we already have the symbol for the property.

Reply:
I wasn't sure of all of the potential symbols `componentProperty` could be since the type of `Member` on `IPropertyReferenceOperation` was `ISymbol` instead of `IPropertySymbol`. Since `IsParameter` expects a property symbol I made an alternate API that just looks at attributes.